wallet/src/components/IbcTransfer.tsx

```diff
@@ -0,0 +1,497 @@
+
+const useRemoteChainAccount = (brandPetname?: PetnameType) => {
+  const ibcAsset =
+    typeof brandPetname === 'string' ? ibcAssets[brandPetname] : undefined;
+
+  const [remoteChainAddress, setRemoteChainAddress] = useState('');
+  const [remoteChainSigner, setRemoteChainSigner] = useState(null);
+  const [remoteChainBalance, setRemoteChainBalance] = useState<bigint | null>(
+    null,
+  );
+
+  const connectWithKeplr = async () => {
+    assert(ibcAsset);
+    // @ts-expect-error window keys
+    const { keplr } = window;
+    const offlineSigner = await keplr.getOfflineSignerOnlyAmino(
+      ibcAsset.chainInfo.chainId,
+    );
+    const accounts = await offlineSigner.getAccounts();
+    setRemoteChainAddress(accounts[0].address);
+    setRemoteChainSigner(offlineSigner);
+  };
+
+  const isRemoteChainAddressInvalid = useMemo(() => {
+    if (!remoteChainAddress) return true;
+
+    try {
+      const { prefix } = fromBech32(remoteChainAddress);
+      return prefix !== ibcAsset?.chainInfo.addressPrefix;
+    } catch {
+      return true;
+    }
+  }, [remoteChainAddress, ibcAsset]);
+
+  useEffect(() => {
+    setRemoteChainBalance(null);
+    if (isRemoteChainAddressInvalid || !ibcAsset) return;
+
+    let isCancelled = false;
+
+    const loadRemoteChainBalance = async () => {
+      const balances = await queryBankBalances(
+        remoteChainAddress,
+        /* @ts-expect-error rpc string */
+        ibcAsset.chainInfo.rpc,
+      );
+      if (isCancelled) return;
+
+      const balance = balances.find(
+        ({ denom }) => denom === ibcAsset.deposit.denom,
+      );
+
+      setRemoteChainBalance(balance ? BigInt(balance.amount) : 0n);
+    };
+
+    void loadRemoteChainBalance();
+
+    return () => {
+      isCancelled = true;
+    };
+  }, [remoteChainAddress, isRemoteChainAddressInvalid, ibcAsset]);
+
+  return {
+    remoteChainBalance,
+    isRemoteChainAddressInvalid,
+    connectWithKeplr,
+    remoteChainAddress,
+    remoteChainSigner,
+    setRemoteChainAddress: (address: string) => {
+      setRemoteChainAddress(address);
+      setRemoteChainSigner(null);
+    },
+  };
+};
+
+// Exported for testing only.
+export const IbcTransferInternal = ({
+  purse,
+  handleClose,
+  direction,
+  keplrConnection,
+}: Params) => {
+  const ibcAsset =
+    typeof purse?.brandPetname === 'string'
+      ? ibcAssets[purse.brandPetname]
+      : undefined;
+
+  const purseBalance = purse?.currentAmount.value;
+
+  const [inProgress, setInProgress] = useState(false);
+  const [error, setError] = useState('');
+  const [amount, setAmount] = useState('');
+  const [isSnackbarOpen, setIsSnackbarOpen] = useState(false);
+
+  const handleCloseSnackbar = _ => {
+    setIsSnackbarOpen(false);
+  };
+
+  const [snackbarMessage, setSnackbarMessage] = useState(<></>);
+
+  const showSnackbar = (
+    isSuccess: boolean,
+    explorerPath: string,
+    transactionHash: string,
+  ) => {
+    setSnackbarMessage(
```

Comment:
no need to put this into state. `isSnackbarOpen` is a signal for whether to render.
```suggestion
    snackbarMessage = <>
```

Reply:
I'm sorry I don't follow. Are you saying we can use an empty message (maybe `null`) as a signal rather than tracking `isSnackbarOpen` separately?

Reply:
No, the opposite. Since we have the Boolean state for whether to present a message we don't also need the message itself to be in state. Just make it a `let` binding in scope.

Reply:
Ah I see, because we only change the content when we change `isSnackbarOpen`, so it will re-render from that. It just seems a bit hazardous to me in case we ever want the content to change while open, simpler to just keep state in `useState`.

Also, using `null` content to hide the snackbar I realized will make the hide animation look choppy.

Reply:
> content to change while open

Yeah if that's a requirement then the content should be in state. But in that case, no need for the snackbar boolean, right? Show the snackbar if there is a message and not otherwise?

Reply:
Since when you set `isSnackbarOpen` to false it doesn't disappear immediately, it fades out over a short duration. So, while it's fading, the content will disappear, making the layout shift.

Reply:
ah thanks for explaining. the current code makes sense. maybe worth some comments though with this context.

Reply:
Just put all the snackbar stuff in a hook, also added a close button instead of relying on clickaway. Also I combined the state into one `useState` so that we only call `setState` once, since it's from an async event this was causing react to render twice unnecessarily.